File: javascript/day4_part2_2023.js
const fs = require('fs');

class Card {
    constructor(winnings, givens, totalCount) {
        this.winnings = winnings;
        this.givens = givens;
        this.totalCount = totalCount;
    }
}

function getPointsForCard(card) {
    let points = 0;
    for (let given in card.givens) {
        if (card.winnings.hasOwnProperty(given)) {
            points += card.givens[given] * card.winnings[given];
        }
    }
    return points;
}

function lexLineIntoCard(line) {
    const [_, cardDataStr, __] = line.split(": ");
    const cardData = cardDataStr.split(" | ");

    const re = /[0-9]{1,2}/g;

    const winnings = {};
    cardData[0].match(re).forEach(point => {
        winnings[point] = (winnings[point] || 0) + 1;
    });

    const givens = {};
    cardData[1].match(re).forEach(point => {
        givens[point] = (givens[point] || 0) + 1;
    });

    return new Card(winnings, givens, 1);
}

fs.readFile('input.txt', 'utf8', (err, data) => {
    if (err) {
        console.error("Error reading file:", err);
        return;
    }

    const input = data.trim();
    const cards = [];

    input.split("\n").forEach(line => {
        if (line.length === 0) {
            return;
        }
        const card = lexLineIntoCard(line);
        cards.push(card);
    });

    cards.forEach((card, i) => {
        const points = getPointsForCard(card);

        for (let j = 1; j <= points; j++) {
            if (cards[i + j]) {
                cards[i + j].totalCount += 1 * card.totalCount;
            }
        }
    });

    let totalCards = 0;
    cards.forEach(card => {
        totalCards += card.totalCount;
    });

    console.log(totalCards);
});

File: go/day4_part2_2023.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strings"
)

type Card struct {
	winnings   map[string]int
	givens     map[string]int
	totalCount int
}

func getPointsForCard(card Card) int {
	points := 0
	for given, count := range card.givens {
		if winningCount, ok := card.winnings[given]; ok {
			points += count * winningCount
		}
	}
	return points
}

func lexLineIntoCard(line string) Card {
	_, cardDataStr, _ := strings.Cut(line, ": ")
	cardData := strings.Split(cardDataStr, " | ")

	re := regexp.MustCompile("[0-9]{1,2}")

	winnings := make(map[string]int)
	for _, point := range re.FindAllString(cardData[0], -1) {
		winnings[point]++
	}

	givens := make(map[string]int)
	for _, point := range re.FindAllString(cardData[1], -1) {
		givens[point]++
	}

	return Card{
		winnings:   winnings,
		givens:     givens,
		totalCount: 1,
	}
}

func main() {
	file, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	input := strings.TrimSpace(string(file))

	var cards []Card

	for _, line := range strings.Split(input, "\n") {
		if len(line) == 0 {
			continue
		}
		card := lexLineIntoCard(line)
		cards = append(cards, card)
	}

	for i, card := range cards {
		points := getPointsForCard(card)

		for j := 1; j <= points; j++ {
			cards[i+j].totalCount += 1 * cards[i].totalCount
		}
	}

	totalCards := 0
	for _, card := range cards {
		totalCards += card.totalCount
	}

	fmt.Println(totalCards)
}